lib/packwerk/package_paths.rb

```diff
@@ -0,0 +1,71 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "pathname"
+require "bundler"
+
+module Packwerk
+  class PackagePaths
+    PACKAGE_CONFIG_FILENAME = "package.yml"
+    PathSpec = T.type_alias { T.any(String, T::Array[String]) }
+
+    extend T::Sig
+    extend T::Generic
+
+    sig do
+      params(
+        root_path: String,
+        package_pathspec: T.nilable(PathSpec),
+        exclude_pathspec: T.nilable(PathSpec),
+        scan_for_packages_outside_of_app_dir: T.nilable(T::Boolean)
+      ).void
+    end
+    def initialize(root_path, package_pathspec, exclude_pathspec = nil, scan_for_packages_outside_of_app_dir = false)
+      @root_path = root_path
+      @package_pathspec = package_pathspec
+      @exclude_pathspec = exclude_pathspec
+      @scan_for_packages_outside_of_app_dir = scan_for_packages_outside_of_app_dir
+    end
+
+    sig {returns(T::Array[Pathname])}
+    def all_paths
+      exclude_pathspec = Array(@exclude_pathspec).dup
+        .push(Bundler.bundle_path.join("**").to_s)
+        .map { |glob| File.expand_path(glob) }
+
+      paths_to_scan = if @scan_for_packages_outside_of_app_dir
+        engine_paths_to_scan.push(@root_path)
+      else
+        [@root_path]
+      end
+
+      glob_patterns = paths_to_scan.product(Array(@package_pathspec)).map do |path, pathspec|
+        File.join(path, pathspec, PACKAGE_CONFIG_FILENAME)
+      end
+
+      Dir.glob(glob_patterns)
+        .map { |path| Pathname.new(path).cleanpath }
+        .reject { |path| exclude_path?(exclude_pathspec, path) }
+    end
+
+    private
+
+    sig { params(globs: T::Array[String], path: Pathname).returns(T::Boolean) }
+    def exclude_path?(globs, path)
+      globs.any? do |glob|
+        path.realpath.fnmatch(glob, File::FNM_EXTGLOB)
+      end
+    end
+
+    sig { returns(T::Array[String]) }
+    def engine_paths_to_scan
+      bundle_path_match = Bundler.bundle_path.join("**")
+
+      Rails.application.railties
```

Comment:
> We don't depend on railties in the gemspec so I'd rather not depend on rails constants here

I'll address this in your second comment below about paths because I think they're kind of related

> I think this also assumes all local engines are packages

That's not exactly accurate. It assumes any local engine might contain packages, but not that the engine itself is a package. The engine doesn't have to contain packages either. It's just adding all the paths to the list of paths that later get scanned for via glob for a `package.yml` file.